common/djangoapps/entitlements/api/v1/views.py

```diff
@@ -140,24 +140,20 @@ def create(self, request, *args, **kwargs):
             course_runs = get_course_runs_for_course(entitlement.course_uuid)
 
             # check if the user has enrollments for any of the course_runs
-            user_run_enrollments = [
-                CourseEnrollment.get_enrollment(user, CourseKey.from_string(course_run.get('key')))
-                for course_run
-                in course_runs
-                if CourseEnrollment.get_enrollment(user, CourseKey.from_string(course_run.get('key')))
-            ]
-
-            # filter to just enrollments that can be upgraded.
-            upgradeable_enrollments = [
-                enrollment
-                for enrollment
-                in user_run_enrollments
-                if enrollment.is_active and enrollment.upgrade_deadline and enrollment.upgrade_deadline > timezone.now()
-            ]
-
-            # if there is only one upgradeable enrollment, convert it from audit to the entitlement.mode
+            upgradeable_enrollments = []
```

Comment:
Could move the below lines into a function and then set upgradeable_enrollments directly

``` python
upgradeable_enrollments = get_upgradeable_enrollments_for_entitlement(entitlement)
```
with 
``` python
def get_upgradeable_enrollments_for_entitlement(entitlement):
  user = entitlement.user
  course_runs = get_course_runs_for_course(entitlement.course_uuid)
  for course_run in course_runs:
    course_run_id = CourseKey.from_string(course_run.get('key'))
    enrollment = CourseEnrollment.get_enrollment(user, course_run_id)
    if (enrollment and
      enrollment.is_active and
      is_course_run_entitlement_fulfillable(course_run_id, entitlement)):
      upgradeable_enrollments.append(enrollment)
```

Reply:
The one line you'd be missing in the above refactoring is initializing an empty list onto which to append enrollment objects.
But I do agree with extracting out the code chunk into its own module.